pkg/acsengine/pki.go

```diff
@@ -87,22 +92,74 @@ func CreatePki(extraFQDNs []string, extraIPs []net.IP, clusterDomain string, caP
 		errors <- err
 	}()
 
+	go func() {
+		var err error
+		organization := make([]string, 1)
+		organization[0] = "system:masters"
+		ip := net.ParseIP("127.0.0.1").To4()
+		peerIPs := append(extraIPs, ip)
+		etcdServerCertificate, etcdServerPrivateKey, err = createCertificate("etcd", caCertificate, caPrivateKey, false, nil, peerIPs, organization)
+		errors <- err
+	}()
+
+	go func() {
+		var err error
+		organization := make([]string, 1)
+		organization[0] = "system:masters"
+		ip := net.ParseIP("127.0.0.1").To4()
+		peerIPs := append(extraIPs, ip)
+		etcdClientCertificate, etcdClientPrivateKey, err = createCertificate("etcd", caCertificate, caPrivateKey, false, nil, peerIPs, organization)
+		errors <- err
+	}()
+
+	for i := 0; i < masterCount; i++ {
+		go func() {
+			var err error
+			organization := make([]string, 1)
+			organization[0] = "system:masters"
+			ip := net.ParseIP("127.0.0.1").To4()
+			peerIPs := append(extraIPs, ip)
+			etcdPeerCertificate := new(x509.Certificate)
+			etcdPeerPrivateKey := new(rsa.PrivateKey)
+			etcdPeerCertificate, etcdPeerPrivateKey, err = createCertificate("etcd", caCertificate, caPrivateKey, false, nil, peerIPs, organization)
+			etcdPeerCertPairs = append(etcdPeerCertPairs, &PkiKeyCertPair{CertificatePem: string(certificateToPem(etcdPeerCertificate.Raw)), PrivateKeyPem: string(privateKeyToPem(etcdPeerPrivateKey))})
```

Comment:
Will this cause race condition since func() are run in parallel?

Reply:
Yes and no. It does mean that the order of the certs is non deterministic. However, I don't think that that's an issue since they all have the same properties and in the end we want 1/3/5 certs signed by the same CA. At this stage it doesn't matter which cert gets assigned to which peer. Please let me know if I am missing something or something is wrong with that logic.

Reply:
In the append function, if two threads try to modify etcdpeerCertPairs at the same time, would one overwrite another?

Reply:
I think I fixed it, thank you for your comment 🙂